### PR TITLE
feat(app): Get protocolDisplayData based on protocol schema

### LINKED ISSUE #3494: Run App: Metadata fields in RA uses the file name instead of metadata key from file
## overview
Run App should support new metadata shape for v3 protocol designer protocols.

## current behavior
- If you change the "name" of the protocol in metadata nothing happens with a v3 protocol, only when you change file name does the Protocol Name change.

## expected behavior
- The metadata key in v3 should be used for the Protocol Name change, and fall back to file name if metadata cannot be found.

Can use [this protocol](https://github.com/Opentrons/opentrons/files/3208408/full-forSethBot.json.zip) to test

### PR DESCRIPTION
**Pull request by Kadee80**
## overview

This PR checks for protocol name, last modified, appName, and appVersion for based on protocol schema versions.

closes #3494

## changelog

- feat(app): Get protocol name for all V3 PD protocols

## review requests

Labeled this a WIP for now - this issue is a little larger than expected as it touches more than just the name field once we start getting schema version from shared data. It also uncovered the lack of v3 protocol file flow type.

[V2 protocol to test](https://github.com/Opentrons/opentrons/files/3262659/Katie.s.Protocol.json.2.zip)

[V3 protocol to test](https://github.com/Opentrons/opentrons/files/3262658/full-forSethBot.json.zip)

- [x] V2 protocols still show correct file info
- [x] V3 protocols show correct file info
- [x] Python protocols show correct file info

### CODE REVIEW

File: app/src/protocol/index.js [hunk 1]
```diff
@@ -108,6 +108,7 @@ type CreatorAppSelector = OutputSelector<
 >
 
 const getName: StringGetter = getter('metadata.protocol-name')
+const getV3Name: StringGetter = getter('metadata.protocolName')
```
Comment (Kadee80) [inline app/src/protocol/index.js]: Kind of at a loss for what to name this - since we need old and new for now and V3 will become the standard soon?

Comment (IanLondon) [reply app/src/protocol/index.js]: We're not dropping support for V1/V2 anytime soon AFAIK, and there will be future versions to co-support. Maybe something vaguely like:

```js
const protocolV1V2GetterPaths = {
  name: 'metadata.protocol-name'
  ...etc
}

const PROTOCOL_GETTER_PATHS_BY_SCHEMA = {
  "1": protocolV1V2GetterPaths,
  "2": protocolV1V2GetterPaths,
  "3": {
    name: 'metadata.protocolName',
    ...etc
  }
}

const getProtocolSchemaVersion = getter('protocol-schema') || getter('schemaVersion') || null
// ...

const pathsForProtocol = getProtocolSchemaVersion(protocol)
const getName = getter(pathsForProtocol.name)
```

File: app/src/protocol/index.js [hunk 2]
```diff
@@ -128,15 +152,50 @@ export const getProtocolFilename: StringSelector = createSelector(
...
+  getProtocolFilename,
+  (data, name) => {
+    if (!data)
+      return {
+        protocolName: name && stripDirAndExtension(name),
+        lastModified: null,
+        appName: null,
+        appVersion: null,
+      }
+    const version = (data && getProtocolSchemaVersion(data)) || 1
```
Comment (IanLondon) [inline app/src/protocol/index.js]: I kinda feel like this `|| 1` might at best case give you the right display data for an invalid json protocol and worst case would make a bug about this just a little harder to reason thru. If `getProtocolSchemaVersion` returns nullsy, then the simplest thing would be to do the same as just above this line, in the `if (!data)` block (return all nulls and name from filename)

File: app/src/protocol/types.js [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
-import type { SchemaV1ProtocolFile } from '@opentrons/shared-data'
+import type {
+  SchemaV1ProtocolFile,
+  SchemaV3ProtocolFile,
+} from '@opentrons/shared-data'
 
 // data may be a full JSON protocol or just a metadata dict from Python
 export type ProtocolData =
   | SchemaV1ProtocolFile<{}>
+  | SchemaV3ProtocolFile<{}>
```
Comment (IanLondon) [inline app/src/protocol/types.js]: to be slightly more safe maybe this should be `SchemaV3ProtocolFile<undefined>` or maybe `SchemaV3ProtocolFile<{||}>`? Run App should never look in `designerApplication.data`

File: shared-data/protocol/index.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+type ProtocolData =
+  | $Shape<SchemaV1ProtocolFile<{}>>
+  | $Shape<SchemaV3ProtocolFile<{}>>
+
+// $FlowFixMe: (ka, 2019-06-10): cant differentiate which file schema file is needed
+export function getProtocolSchemaVersion(data: ProtocolData): ?number {
+  if (data.schemaVersion) {
+    return data.schemaVersion
+  } else if (data['protocol-schema']) {
```
Comment (IanLondon) [inline shared-data/protocol/index.js]: To be stricter about the "legacy" `protocol-schema` key, we could do very explicit `if (data['protocol-schema'] == '1.0.0') return 1; else if (data['protocol-schema'] == '2.0.0') return 2`. If it's `'3lolz'` or even if it's `3.0.0` or `3`, those are invalid ways to designate a protocol as v3

I know this strictness is totally overkill for how this fn is used in this PR, but we're going to use this fn for protocol migration very soon and for that I'm not comfortable with it being this permissive

File: app/src/protocol/types.js (hunk 3 above)
Comment (Kadee80) [reply app/src/protocol/types.js]: This won't work for Flow unfortunately

File: app/src/protocol/index.js (hunk 2 above)
Comment (Kadee80) [reply app/src/protocol/index.js]: This is needed because python protocols return metadata formatted like a V1 protocol to the app

**Pull request merged**